lib/zip/extra_field/universal_time.rb

```diff
@@ -8,20 +8,52 @@ def initialize(binstr = nil)
       @ctime = nil
       @mtime = nil
       @atime = nil
-      @flag  = nil
-      binstr && merge(binstr)
+      @flag  = 0
+
+      merge(binstr) unless binstr.nil?
+    end
+
+    attr_reader :atime, :ctime, :mtime, :flag
+
+    def atime=(time)
+      time.nil? ? @flag &= 0b101 : @flag |= 0b010
+      @atime = time
     end
 
-    attr_accessor :atime, :ctime, :mtime, :flag
+    def ctime=(time)
+      time.nil? ? @flag &= 0b011 : @flag |= 0b100
+      @ctime = time
+    end
+
+    def mtime=(time)
+      time.nil? ? @flag &= 0b110 : @flag |= 0b001
+      @mtime = time
+    end
 
     def merge(binstr)
       return if binstr.empty?
+
       size, content = initial_parse(binstr)
-      size || return
-      @flag, mtime, atime, ctime = content.unpack('CVVV')
-      mtime && @mtime ||= ::Zip::DOSTime.at(mtime)
-      atime && @atime ||= ::Zip::DOSTime.at(atime)
-      ctime && @ctime ||= ::Zip::DOSTime.at(ctime)
+      return if size.nil? || size <= 0
+
+      flags, *times = content.unpack('Cl<l<l<')
+      @flag = flags
+
+      # Parse the timestamps, in order, based on which flag is set.
+      times.each do |time|
```

Comment:
I think your reading of the spec is correct --- good find. I think it may be (marginally) clearer without the loop, if this also looks right to you?
```rb
@flag, *times = content.unpack('Cl<3')
return if times[0].nil?
@mtime ||= ::Zip::DOSTime.at(times.shift) unless @flag & MTIME_MASK == 0
return if times[0].nil?
@atime ||= ::Zip::DOSTime.at(times.shift) unless @flag & ATIME_MASK == 0
return if times[0].nil?
@ctime ||= ::Zip::DOSTime.at(times.shift) unless @flag & CTIME_MASK == 0
```

Reply:
This is somewhat neater and doesn't need a second copy of the flags. I'll go with this.